003-graphql.md

```diff
@@ -0,0 +1,42 @@
+ - We reimplemented cross-application includes (such as `?include=user` on a list of posts in Rogue) in multiple apps. This caused different backend services to have to act as a "gateway" depending on where a request started.
+ - We found that we would eventually implement the same batching & caching logic in each application (for example, fetching campaigns [in Rogue](https://git.io/fjjGZ) or users [in our now-defunct competitions app](https://github.com/DoSomething/gladiator/blob/534ca9834a5369ededaed56deff33404bbc94e99/app/Repositories/DatabaseUserRepository.php#L62-L163)). In each case, we'd have to reconsider the right order-of-operations and maximum batch size for good performance.
+ - Because each application maintained it's own caches, it was hard to maintain a consistent cache policy. (For example, one application might cache users for an hour while another didn't cache them at all).
+ - As our schema grew in each application, we became wary of adding new fields that would bloat responses. This slowed down development & made it harder to justify building shorter-term features or experiments.
+ - Finally, we had to do a lot of busy-work to [normalize our Redux state](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape) in Phoenix, which slowed down development of new features. (See [DoSomething/phoenix-next#767](https://github.com/DoSomething/phoenix-next/pull/767) and [DoSomething/phoenix-next#1123](https://github.com/DoSomething/phoenix-next/pull/1123)).
+
+## Proposal
+
+We built our GraphQL gateway (see [graphql.dosomething.org](https://graphql.dosomething.org/)) in November 2017. We have gradually added more resources (such as Contentful spaces and embed metadata) and applications (such as Gambit) over time.
+
+This gives us a single [universal schema](https://principledgraphql.com/integrity#1-one-graph) (so developers don't need to worry about which application holds which data). It lets us optimally [batch requests to backend services](https://www.apollographql.com/docs/graphql-tools/connectors/#dataloader-and-caching) and acts as a centralized "wrapper" around third-party APIs like [Contentful](https://www.contentful.com/).
+
+It also allows us to organically evolve our API over time (since we can add fields without bloating existing requests that don't need them, and [deprecate old fields](https://www.apollographql.com/docs/graphql-tools/generate-schema/#descriptions--deprecations) without breaking existing clients). We also get fantastic developer tools, like the [Apollo Chrome extension](https://chrome.google.com/webstore/detail/apollo-client-developer-t/jdkknkkbebbapilgoeccciglkfbmbnfm?hl=en-US) and [Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo).
+
```

Comment:
I keep coming back to "Does this make a clear case for why we've been standardizing data fetching with GraphQL?" and thinking of more, and it's related to this section -- but it's pretty great that we have / will-have-once-everything-is-ported-over standard camelCase naming standard on all resources. Related to deprecating old fields is having easier control of the names of resources, e.g. many of the Gambit schema names don't match the Gambini content types, and although the neat freak in all of us would prefer the name to be same all over, it's questionable whether a migration to change the machine name is worth it (especially if we want to change it again and again)

Reply:
We also get built in documentation by using the Playground (again, maybe a given - but definitely helps with onboarding I think)

Reply:
Oh man, yeah the playground & its built-in documentation is one of my absolute favorite GraphQL features! Definitely worth a shout-out in here.